Enable context-flattening to support context field searching

nickmarden opened the pull request:
Description

  Resolves open-telemetry/opentelemetry-php#1837

  This PR adds an optional configuration to flatten log context attributes into individual OTLP attributes, improving searchability in observability backends.

  Problem

  The Laravel LogWatcher currently JSON-encodes the entire log context into a single context attribute:

  context: {"http":{"method":"GET","path":"/users","status":200},"user_id":"123"}

  This makes it impossible to search/filter by specific values (e.g., http.method = "GET") in backends like SigNoz, Jaeger, or Grafana.

  Solution

  A new configuration option OTEL_PHP_LARAVEL_LOG_ATTRIBUTES_FLATTEN that, when enabled, spreads context as individual attributes with dot-notation for nested arrays:

  http.method: GET
  http.path: /users
  http.status: 200
  user_id: 123

  Changes

  - src/Watchers/LogWatcher.php: Added flattening logic with ConfigurationResolver, dot-notation for nested arrays, and proper handling of Stringable objects
  - README.md: Documented the new configuration option with examples
  - tests/Unit/Watchers/LogWatcherTest.php: Added 11 unit tests covering default behavior, flattened mode, nested arrays, exceptions, and edge cases

  Backward Compatibility

  - Default behavior unchanged (off by default)
  - Existing tests continue to pass

  Testing

  ./vendor/bin/phpunit --testsuite=unit      # 18 tests, 42 assertions
  ./vendor/bin/phpunit --testsuite=integration  # 21 tests, 120 assertions
  
  Screenshot
  
  
<img width="766" height="727" alt="Screenshot 2025-12-17 at 6 56 35 PM" src="https://github.com/user-attachments/assets/4343b62c-0151-4153-be8b-0bdb428d9580" />

nickmarden committed:
Enable context-flattening to support context field searching

nickmarden committed:
fix(tests): Remove deprecated EXCEPTION_ESCAPED attribute

nickmarden committed:
fix(tests): linting fix for new_with_parentheses

ChrisLightfootWild commented on src/Instrumentation/Laravel/src/Watchers/LogWatcher.php:
```diff
@@ -56,20 +68,28 @@ public function recordLog(MessageLogged $log): void
+            ->setSeverityNumber(Severity::fromPsr3($log->level));
+
+        if ($this->flattenAttributes) {
+            foreach ($this->buildFlattenedAttributes($contextToProcess) as $key => $value) {
```
Could https://laravel.com/docs/13.x/helpers#method-array-dot work here instead? 🤔

arthurmartelli replied:
The implementation is much cleaner with that array method :)

arthurmartelli committed:
refactor(log-watcher): flatten context via Arr::dot()

Replace the custom recursive flattenArray() method and the
manual normalization loop with a single array_map() +
Arr::dot() expression, removing ~25 lines of bespoke logic.

ChrisLightfootWild commented:
Hey @nickmarden / @arthurmartelli - firstly, thanks for your contribution here.

It appears that I made an error in merging this though, as it has conflicted with behaviour that changed in a later PR #511 (which had already been merged ahead of this one).

This may have to be reverted to restore the previous behaviour - however, I think it is something that perhaps this PR was trying to achieve in the first instance.

Is there any additional behaviour that this PR provides in addition to the changes introduced in #511 which you were still hoping to accomplish?

nickmarden commented:
@ChrisLightfootWild lemme take a look to see if #511 accomplishes what we need.

nickmarden commented:
Thanks @ChrisLightfootWild  for bringing this to our attention.

After reviewing #511, I think it solves the JSON-encoding problem by passing context as a structured array, which is better, but it doesn't solve the faceted search use case in systems like SigNoz. SigNoz (and similar backends) require individual top-level OTLP attributes to support filtering/faceting. A single context attribute containing a nested array/kvlist is not decomposed into searchable facets.

The value proposition of #491 is the opt-in `OTEL_PHP_LARAVEL_LOG_ATTRIBUTES_FLATTEN` mode, which spreads context fields as individual dot-notation OTLP attributes (e.g., `http.method` = GET, `user_id` = 123). That's what enables faceted log search.

The issue with the current merged state is that the non-flatten path in #491 regressed #511's improvement by reverting to json_encode() instead of keeping the structured array. A fix could be:

- Non-flatten default: keep #511's structured array (`setAttribute('context', $context)`)
- Flatten mode: individual dot-notation attributes as #491 implemented

If you'd like, @arthurmartelli or I can open a follow-up PR that corrects the non-flatten regression while preserving the (optional) flatten feature.

nickmarden commented:
@ChrisLightfootWild please have a look at #460 - it keeps the desired context-flattening while preserving the #511 structured array behavior as default.
